### ISSUE TITLE
Documentation about middlewares seems to be incorrect

### ISSUE THREAD

**Issue by lvyyljn** [ThreeMammals/Ocelot#1686]
I'm currently working on updating the response from downstream and noticed some inconsistency
As it's stated here https://ocelot.readthedocs.io/en/latest/features/middlewareinjection.html
_Obviously you can just add middleware as normal before the call to app.UseOcelot() It cannot be added after as Ocelot does not call the next middleware._

And it's not true, as the last middleware in the **BuildOcelotPipeline** method is **HttpRequesterMiddleware** that calls the next middleware, I think documentation should be changed

## Specifications

  - Version: 18.0.0
  - Platform: .NET 6.0

**Comment by raman-m** [comment]
Hi lvyyljn!
Thanks for your interest in Ocelot!

> ## Specifications
>  - Version: **18.0.0**
>  - Platform: **.NET 6.0**

Why do you use previous version? Your project is based on .NET 6 ?
It is time to upgrade to .NET 7 and Ocelot 19.0.2.

**Comment by lvyyljn** [comment]
Hi @raman-m, it was predefined before me to use .NET 6, so we have to use 18.0.0, would be glad to update but still)
For the 19.0.2 documentation to looks outdated

**Comment by raman-m** [comment]
What is your real name?
Where are you from?
Why is your GitHub profile empty?

**Comment by lvyyljn** [comment]
Is this relevant?)
I am only concerned about documentation being outdated,
My name is Denys and I am from Ukraine, my profile is empty as it's first time I need to create issue here

**Comment by lvyyljn** [comment]
> It says you cannot configure the pipeline after Ocelot's builder has ran.

You mean here the Ocelot's pipeline correct?
And maybe better would be 

> Next called middlewares won't affect Ocelot configuration

Maybe I misunderstood something?

**Comment by raman-m** [comment]
@lvyyljn commented [on Aug 28](#issue-1870099151):
> And it's not true, as the last middleware in the **BuildOcelotPipeline** method is **HttpRequesterMiddleware** that calls the next middleware, I think documentation should be changed

Why would you like to discuss the internal [HttpRequesterMiddleware](../blob/develop/src/Ocelot/Requester/Middleware/HttpRequesterMiddleware.cs) ? It is a part of pipeline but it is private and cannot be overridden because this middleware is not user's one!
From other side, yes, this is [the last middleware](../blob/develop/src/Ocelot/Middleware/OcelotPipelineExtensions.cs#L149) in overall Ocelot & ASP.NET pipeline, but it serves non-user operation.
The last user middleware which can be overridden is: [PreQueryStringBuilderMiddleware](https://github.com/ThreeMammals/Ocelot/blob/develop/src/Ocelot/Middleware/OcelotPipelineExtensions.cs#L131) being read from pipeline configuration object (`pipelineConfiguration`).
In general the docs says that [PreQueryStringBuilderMiddleware](../blob/develop/docs/features/middlewareinjection.rst?plain=1#L38) is the last **user middleware**! And there are no another user middlewares in pipeline.

It seems you were confused in differentiation of system (private) and user (public) middleware. and yes, docs can be corrected somehow to eliminate this confusion.

**Comment by raman-m** [comment]
@lvyyljn commented [on 2023-08-29 at 08:14](#issuecomment-1696977292):

> Maybe I misunderstood something?

Yes, you did.

---
> And maybe better would be "**Next called middlewares won't affect Ocelot configuration**"

You are welcome to open PR!

**Comment by lvyyljn** [comment]
@raman-m Totally agree with you on the private and public, and I don't intend to override any of the private middlewares
I wanted to add middleware after UseOcelot, and was confused as the documentation stated Ocelot doesn't call the next middleware

**Comment by lvyyljn** [comment]
Yes, you're right `UseOcelot` builds a pipeline, but the build method underhood doesn't close to the ability to edit the pipeline further, it's just a checkpoint, You can try it on a local machine within a few minutes, this code works fine without any error
```
app.UseSwaggerForOcelotUI(opt =>
{
    opt.PathToSwaggerGenerator = "/swagger/docs";
    opt.ReConfigureUpstreamSwaggerJson = AlterUpstream.AlterUpstreamSwaggerJson;
}).UseOcelot().Wait();

app.UseMiddleware<RequestCultureMiddleware>();
```
And no, I won't create a PR for docs

**Comment by raman-m** [comment]
> @lvyyljn commented [last week](#issuecomment-1697472951)

👌 You've investigated that a custom middleware can be attached after building pipeline. Great! But it is configuration stage during app startup.

---
> And no, I won't create a PR for docs

Can you close the issue please?
I don't see any benefits from further conversation.

You can do any Ocelot research you like.

**Comment by raman-m** [comment]
Duplicate of #1678

**Comment by raman-m** [comment]
@lvyyljn commented [on Aug 29](#issuecomment-1697472951)
> And no, I won't create a PR for docs

**FYI**
**Your issue-question will be closed by #1678** 
I have added all results of this discussion to mentioned PR. See the commit [3cd13d1](../pull/1678/commits/3cd13d1d86a6da348f8e3af9a51b5e82c53ca800)
